### PR TITLE
Feature/dynamic asset pool

### PR DESCRIPTION
**Pull request by salmanbao**
# Description - 
 * This PR adds a proxy to the master code for cloning the smart contract. 
 * Changes the Openzeppeling library to Openzeppeling-upgradable.
 * add an extra function to receive the ETH from the other's account.
 * create Create Token function to clone the ERC20 token.
 * create an initializer to initialize the default value of the  ERC20 contract.
 * override the decimal value in the ERC20Upgradeable contract.
 * write test for assetPoolUpgradeable.

### CODE REVIEW

File: contracts/AssetPool.sol [hunk 1]
```diff
@@ -1,41 +1,56 @@
...
+    assetToLp[assetAddress] = address(token);
+  }
+  // Proxy code ends here
 
   function depositAsset(uint256 depositAmount, address assetAddress) external payable nonReentrant {
     require(depositAmount > 0, "1: Deposit Amount cannot be 0");
-    require(allowedAssets[assetAddress] == true, "Asset not approved for deposit");
-    userInfo[msg.sender][assetAddress] += (depositAmount);
-    IERC20 depositToken = IERC20(assetAddress);
+    userInfo[msg.sender][assetAddress] += depositAmount;
```
Comment (Akhilleshgoswami) [inline contracts/AssetPool.sol:39]: it is a good idea to add the money into user's account once the tx is done.

File: contracts/AssetPool.sol [hunk 2]
```diff
@@ -1,41 +1,56 @@
...
     emit Provide(msg.sender, depositAmount);
   }
 
-  function withdrawAsset(uint crAmount, address assetAddress) external payable nonReentrant {
-    require(crAmount <= balanceOf(msg.sender), "1: Amount is too large");
-    require(crAmount > 0, "1: Amount cannot be zero");
-    require(allowedAssets[assetAddress] == true, "Asset not approved for withdrawal");
+  function withdrawAsset(uint256 assetsValue, address assetAddress) external payable nonReentrant {
+    require(assetsValue > 0, "1: Amount cannot be zero");
     require(userInfo[msg.sender][assetAddress] > 0, "1: Withdrawal balance is 0");
```
Comment (Akhilleshgoswami) [inline contracts/AssetPool.sol:48]: can remove this condition and can call the balance of function from the contract and check the user balance  that would be great.

File: contracts/AssetPool.sol [hunk 3]
```diff
@@ -1,41 +1,56 @@
...
+import "@openzeppelin/contracts/utils/math/SafeMath.sol";
+import "hardhat/console.sol";
+import "./Proxy.sol";
+import "@openzeppelin/contracts/security/ReentrancyGuard.sol";
 
-contract AssetPool is LiquidityPoolInterfaces, Ownable, ReentrancyGuard, ERC20("Cruize Protection LP Token", "cr") {
+contract AssetPool is Proxy, Ownable, ReentrancyGuard {
   using SafeMath for uint256;
-  mapping(address => mapping(address => uint)) public userInfo;
+  mapping(address => mapping(address => uint256)) public userInfo;
```
Comment (Akhilleshgoswami) [inline contracts/AssetPool.sol:11]: can remove. we can call the contract balance of function in to withdraw function.

File: contracts/AssetPool.sol [hunk 4]
```diff
@@ -1,41 +1,56 @@
...
   mapping(address => bool) private allowedAssets;
```
Comment (Akhilleshgoswami) [inline contracts/AssetPool.sol:12]: can be removed.

File: contracts/AssetPool.sol [hunk 5]
```diff
@@ -1,41 +1,56 @@
...
+  function withdrawAsset(uint256 assetsValue, address assetAddress) external payable nonReentrant {
+    require(assetsValue > 0, "1: Amount cannot be zero");
     require(userInfo[msg.sender][assetAddress] > 0, "1: Withdrawal balance is 0");
-    _burn(msg.sender, crAmount);
-    userInfo[msg.sender][assetAddress] -= crAmount;
-    IERC20 withdrawToken = IERC20(assetAddress);
-    require(withdrawToken.transfer(msg.sender, crAmount), "1: Transfer failed!");
-    emit Withdraw(msg.sender, crAmount);
+    ILPtoken(assetToLp[assetAddress]).burn(msg.sender,assetsValue);
+    userInfo[msg.sender][assetAddress] -= assetsValue;
```
Comment (Akhilleshgoswami) [inline contracts/AssetPool.sol:50]: can remove. balanceOf()  can call here

File: contracts/AssetPool.sol [hunk 6]
```diff
@@ -1,41 +1,56 @@
 // SPDX-License-Identifier: MIT
-pragma solidity >=0.4.22 <0.9.0;
+pragma solidity ^0.8.0;
 import "./LiquidityPoolInterfaces.sol";
+import "@openzeppelin/contracts/utils/math/SafeMath.sol";
```
Comment (Akhilleshgoswami) [inline contracts/AssetPool.sol:4]: are we using this if not we can remove this import?

File: contracts/AssetPool.sol [hunk 7]
```diff
@@ -1,41 +1,56 @@
...
+  event Provide(address indexed account,uint256 writeAmount);
+
+  // This is proxy code
+  address immutable masterContract;
+
+  constructor(address _masterContract) {
+    masterContract = _masterContract;
+  }
+
+  function createToken(string memory name_, string memory symbol_, address assetAddress) external onlyOwner nonReentrant {
```
Comment (Akhilleshgoswami) [inline contracts/AssetPool.sol:30]: follow the same name format for the parameters name for all external functions.

File: contracts/AssetPool.sol [hunk 8]
```diff
@@ -1,41 +1,56 @@
...
+      address indexed account,
+      uint256 writeAmount
+  );
+  event Provide(address indexed account,uint256 writeAmount);
+
+  // This is proxy code
+  address immutable masterContract;
+
+  constructor(address _masterContract) {
+    masterContract = _masterContract;
```
Comment (Akhilleshgoswami) [inline contracts/AssetPool.sol:27]: you can change the name as CrContract.

File: contracts/AssetPool.sol [hunk 9]
```diff
@@ -1,41 +1,56 @@
...
+  }
+
+  function createToken(string memory name_, string memory symbol_, address assetAddress) external onlyOwner nonReentrant {
+    CruizeToken token = CruizeToken(createClone(masterContract));
+    token.init(name_, symbol_);
+    assetToLp[assetAddress] = address(token);
+  }
+  // Proxy code ends here
 
   function depositAsset(uint256 depositAmount, address assetAddress) external payable nonReentrant {
```
Comment (Akhilleshgoswami) [inline contracts/AssetPool.sol:37]: asset address can be changed to reserve  , and deposit  amount could be change to amount

File: contracts/AssetPool.sol [hunk 10]
```diff
@@ -1,41 +1,56 @@
...
-    _mint(msg.sender, depositAmount);
+    ILPtoken(assetToLp[assetAddress]).mint(msg.sender,depositAmount);
     emit Provide(msg.sender, depositAmount);
   }
 
-  function withdrawAsset(uint crAmount, address assetAddress) external payable nonReentrant {
-    require(crAmount <= balanceOf(msg.sender), "1: Amount is too large");
-    require(crAmount > 0, "1: Amount cannot be zero");
-    require(allowedAssets[assetAddress] == true, "Asset not approved for withdrawal");
+  function withdrawAsset(uint256 assetsValue, address assetAddress) external payable nonReentrant {
```
Comment (Akhilleshgoswami) [inline contracts/AssetPool.sol:46]: the assetvalue can change to amount and the asset address could be changed with the token.

File: contracts/AssetPool.sol [hunk 11]
```diff
@@ -1,41 +1,56 @@
...
+  }
+  // Proxy code ends here
 
   function depositAsset(uint256 depositAmount, address assetAddress) external payable nonReentrant {
     require(depositAmount > 0, "1: Deposit Amount cannot be 0");
-    require(allowedAssets[assetAddress] == true, "Asset not approved for deposit");
-    userInfo[msg.sender][assetAddress] += (depositAmount);
-    IERC20 depositToken = IERC20(assetAddress);
+    userInfo[msg.sender][assetAddress] += depositAmount;
+    ILPtoken depositToken = ILPtoken(assetAddress);
```
Comment (Akhilleshgoswami) [inline contracts/AssetPool.sol:40]: You have to put a check condition here for the deposit asset that is we are allowed to deposit this asset or not?
for ex. someone says i want to put the link but we. are not allowing anyone to deposit link's that condition must be checked here. before accessing the user's assets

File: contracts/AssetPool.sol [hunk 12]
```diff
@@ -1,41 +1,56 @@
...
+import "./Proxy.sol";
+import "@openzeppelin/contracts/security/ReentrancyGuard.sol";
 
-contract AssetPool is LiquidityPoolInterfaces, Ownable, ReentrancyGuard, ERC20("Cruize Protection LP Token", "cr") {
+contract AssetPool is Proxy, Ownable, ReentrancyGuard {
   using SafeMath for uint256;
-  mapping(address => mapping(address => uint)) public userInfo;
+  mapping(address => mapping(address => uint256)) public userInfo;
   mapping(address => bool) private allowedAssets;
+  mapping(address => address) public assetToLp;
```
Comment (Akhilleshgoswami) [inline contracts/AssetPool.sol:13]: the name should be specified  the work of the mapping

File: contracts/AssetPool.sol [hunk 13]
```diff
@@ -1,41 +1,56 @@
...
+  address[] public tokenAddress;
+
+
+  event Withdraw(
+      address indexed account,
+      uint256 writeAmount
+  );
+  event Provide(address indexed account,uint256 writeAmount);
```
Comment (Akhilleshgoswami) [inline contracts/AssetPool.sol:21]: the events must follow the same style. Also, the name of the event provided could be changed to deposit. the written amount could be changed to receive a token.

File: contracts/AssetPool.sol [hunk 14]
```diff
@@ -1,41 +1,56 @@
...
 
-contract AssetPool is LiquidityPoolInterfaces, Ownable, ReentrancyGuard, ERC20("Cruize Protection LP Token", "cr") {
+contract AssetPool is Proxy, Ownable, ReentrancyGuard {
   using SafeMath for uint256;
-  mapping(address => mapping(address => uint)) public userInfo;
+  mapping(address => mapping(address => uint256)) public userInfo;
   mapping(address => bool) private allowedAssets;
+  mapping(address => address) public assetToLp;
+  address[] public tokenAddress;
+
```
Comment (Akhilleshgoswami) [inline contracts/AssetPool.sol:15]: can remove these unwanted spaces.

**Review by Akhilleshgoswami — CHANGES_REQUESTED** [review]
I have added some comments. please do change them and feel free to ask me if have any doubts.

File: contracts/AssetPool.sol (hunk 11 above)
Comment (Prithviraj8) [reply contracts/AssetPool.sol:40]: yeah basically line 12 was supposed to do that. We should not have removed it I guess ?

File: contracts/AssetPool.sol (hunk 5 above)
Comment (Prithviraj8) [reply contracts/AssetPool.sol:50]: sorry which line are you asking to be removed here ? 49 or 50 ?

File: contracts/CRToken.sol [hunk 15]
```diff
@@ -0,0 +1,224 @@
...
+
+    string private _name;
+    string private _symbol;
+
+    /**
+     * Mint the cr token against the staked asset
+     * It cannot be called by any one other than the smart contract
+     */
+    function mint(address _to,uint256 _amount)external onlyOwner {
+    _mint(_to,_amount);
```
Comment (Prithviraj8) [inline contracts/CRToken.sol:26]: Shouldn't be an indentation here: A tab or space ? Or was this done automatically by your code formatter?

File: contracts/CRToken.sol [hunk 16]
```diff
@@ -0,0 +1,224 @@
...
+    }
+
+    /**
+     * Burn the cr token after withdrawal of the staked asset
+     * It cannot be called by any one other than the smart contract
+     */
+    function burn(address _to,uint256 _amount)external onlyOwner {
+    _burn(_to,_amount);
```
Comment (Prithviraj8) [inline contracts/CRToken.sol:34]: Shouldn't be an indentation here: A tab or space ? Or was this done automatically by your code formatter?

File: contracts/AssetPool.sol (hunk 11 above)
Comment (Akhilleshgoswami) [reply contracts/AssetPool.sol:40]: we have managed that thing.Karna does not put the updated code

File: contracts/CRToken.sol (hunk 15 above)
Comment (Akhilleshgoswami) [reply contracts/CRToken.sol:26]: yup right

File: contracts/AssetPool.sol (hunk 9 above)
Comment (salmanbao) [reply contracts/AssetPool.sol:37]: why this function is payable, as we are not handling ethers in this smart contract so if someone unintentionally
send the ethers to this function then those ethers will be locked forever

File: contracts/assetsPool/AssetPoolUpgradeable.sol [hunk 17]
```diff
@@ -0,0 +1,154 @@
...
+ * @author CRUIZE.
+ * @title Cruize AssetPool.
+ */
+contract AssetPoolUpgradeable is
+    Proxy,
+    OwnableUpgradeable,
+    ReentrancyGuardUpgradeable
+{
+    mapping(address => address) public lpTokens;
+    // change WETH address when you use different test netwrok CURRENT WETH KOVEN network address.
```
Comment (Prithviraj8) [inline contracts/assetsPool/AssetPoolUpgradeable.sol:19]: Hey what are we trying to say in this comment?
Needs to be phrased better 😅.

File: contracts/assetsPool/AssetPoolUpgradeable.sol [hunk 18]
```diff
@@ -0,0 +1,154 @@
...
+        CRTokenUpgradeable token = CRTokenUpgradeable(createClone(crContract));
+        token.initialize(name, symbol, decimal);
+        lpTokens[tokenAddress] = address(token);
+        emit CreateToken(address(token), name, symbol, decimal);
+    }
+
+    /*    ************* PROXY CODE END ****************  */
+
+    /*
+     * @function  depositeAsset  will deposit user's asset to CRUIZE.
```
Comment (Prithviraj8) [inline contracts/assetsPool/AssetPoolUpgradeable.sol:71]: Can write this:
will deposit user's asset to CRUIZE Asset Pool.

File: contracts/assetsPool/AssetPoolUpgradeable.sol [hunk 19]
```diff
@@ -0,0 +1,154 @@
...
+     * @param     amount          number of token to be deposit.
+     * @param     reserve         token address to deposit.
```
Comment (Prithviraj8) [inline contracts/assetsPool/AssetPoolUpgradeable.sol:73]: Any reason why we're calling it "**reserve**" and not just **token_address** or just **token** ? Just curious, is using **reserve** a common term for liquidity pool contracts in solidity ?

Comment (Prithviraj8) [reply contracts/assetsPool/AssetPoolUpgradeable.sol:73]: And I see that we're already using the word **token** for token addresses in withdraw asset function.

Comment (Akhilleshgoswami) [reply contracts/assetsPool/AssetPoolUpgradeable.sol:73]: yup, so the reserve I saw some of the liquidity contracts like uni swap and Avve they use this naming convention.

File: contracts/assetsPool/AssetPoolUpgradeable.sol (hunk 17 above)
Comment (Akhilleshgoswami) [reply contracts/assetsPool/AssetPoolUpgradeable.sol:19]: lol, 😂  that's called a quick bet.

**Review by salmanbao — COMMENTED** [review]
we should check newly created crTokens ownerships in test/AssetPool.test.js

**Pull request merged**